src/Bicep.Core/SourceCode/SourceArchive.cs

```diff
@@ -150,44 +187,127 @@ public static Stream PackSourcesIntoStream(Uri entrypointFileUri, IReadOnlyDicti
+
                         WriteNewFileEntry(tarWriter, archivePath, source);
-                        filesMetadata.Add(new SourceFileInfoEntry(location, archivePath, kind));
+                        filesMetadata.Add(new SourceFileInfoEntry(relativePath, archivePath, kind));
 
-                        if (file.FileUri == entrypointFileUri)
+                        if (PathHelper.PathComparer.Equals(file.FileUri, entrypointFileUri))
                         {
                             if (entryPointPath is not null)
                             {
                                 throw new ArgumentException($"{nameof(SourceArchive)}.{nameof(PackSourcesIntoStream)}: Multiple source files with the entrypoint \"{entrypointFileUri.AbsoluteUri}\" were passed in.");
                             }
 
-                            entryPointPath = location;
+                            entryPointPath = relativePath;
                         }
+
+                        sourceUriToRelativePathMap.Add(file.FileUri, relativePath);
                     }
 
                     if (entryPointPath is null)
                     {
                         throw new ArgumentException($"{nameof(SourceArchive)}.{nameof(PackSourcesIntoStream)}: No source file with entrypoint \"{entrypointFileUri.AbsoluteUri}\" was passed in.");
                     }
 
-                    // Add the metadata file
-                    var metadataContents = CreateMetadataFileContents(baseFolderUri, entryPointPath, filesMetadata, documentLinks);
+                    // Convert links
+                    var pathBasedLinks = UriDocumentLinksToPathBasedLinks(sourceUriToRelativePathMap, documentLinks);
+
+                    // Create and add the metadata file
+                    var metadataContents = CreateMetadataFileContents(entryPointPath, filesMetadata, pathBasedLinks);
                     WriteNewFileEntry(tarWriter, MetadataFileName, metadataContents);
                 }
             }
 
             stream.Seek(0, SeekOrigin.Begin);
             return stream;
+        }
+
+        private static (string relativePath, string archivePath) CalculateRelativeAndArchivePaths(string path, string root, string rootName)
+        {
+            // Replace root of the path with root's "friendly" name to help avoid unintended
+            //   disclosure of user paths
+            var relativePath = Path.GetRelativePath(root, path);
+            Debug.Assert(!relativePath.StartsWith("../"), $"All paths should be under one of the roots");
+            relativePath = SourceCodePathHelper.NormalizeSlashes($"{rootName}{relativePath}");
+
+            // Handle illegal/problematic file characters in the path we use inside the archive
+            string archivePath = relativePath;
+            foreach (char ch in PathCharsToAvoid)
+            {
+                archivePath = archivePath.Replace(ch, '_');
```

Comment:
[nit] Would be more efficient to make `PathCharsToAvoid` into a hash set, do the replacement in a single iteration over `archivePath`.

Reply:
Depends.  My thinking was that string.Replace() is a NOOP if the string doesn't contain any of the avoided characters, which I figured was the common case.  The loop case requires creating a new string every time with StringBuilder() (unless you do some optimization that's not worth the effort for this :-).

Some quick timings in the unit tests show they average out about the same.

In the end I went with chatGPT's suggestion, which is essentially equivalent to a loop with StringBuilder:

            // Handle illegal/problematic file characters in the path we use inside the archive
            var archivePath = new string(relativePath.Select(ch => PathCharsToAvoid.Contains(ch) ? '_' : ch).ToArray());

Reply:
That was fun (seriously)!

Reply:
BTW, ImmutableHashset was about 20% slower than Hashset, but I obviously still went with it.

Reply:
Speed probably isn't a huge issue here, so not worth changing at this point. A compromise would probably be to use `IReadOnlySet<T>` - that way it would be hard to accidentally mutate the set:

```csharp
 private static readonly IReadOnlySet<char> PathCharsToAvoid = ...
    .ToHashSet();
```